qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java

```diff
@@ -91,68 +113,114 @@ public Result run( String command, Object... args) {
-
+        final Path stdOut;
+        final Path stdErr;
         try {
+            Path tmpDir = Paths.get(System.getProperty("java.io.tmpdir"));
+            Files.createDirectories(tmpDir);
+            stdOut = Files.createTempFile(tmpDir, getClass().getName(), ".out");
+            stdErr = Files.createTempFile(tmpDir, getClass().getName(), ".err");
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
 
-            Process process = builder.start();
-
-            StringBuilder stdout = new StringBuilder();
-            StringBuilder stderr = new StringBuilder();
-
-            Thread stdoutThread = new Thread(new StreamCollector(process.getInputStream(), stdout));
-            Thread stderrThread = new Thread(new StreamCollector(process.getErrorStream(), stderr));
-
-            stdoutThread.start();
-            stderrThread.start();
-
-            stdoutThread.join();
-            stderrThread.join();
-
-            int exitCode = process.waitFor();
+        redirectOutAndErr(builder, stdOut, stdErr);
 
-            return new Result(exitCode, stdout.toString(), stderr.toString());
+        try {
+            Process process = builder.start();
+            if (process.waitFor(10, TimeUnit.MINUTES) == false) {
+                if (process.isAlive()) {
+                    process.destroyForcibly();
+                }
+                Result result = new Result(
+                    -1,
+                    readFileIfExists(stdOut),
+                    readFileIfExists(stdErr)
+                );
+                throw new IllegalStateException(
+                    "Timed out running shell command: " + command + "\n" +
+                    "Result:\n" + result
+                );
+            }
 
-        } catch (IOException | InterruptedException e) {
+            Result result = new Result(
+                process.exitValue(),
+                readFileIfExists(stdOut),
+                readFileIfExists(stdErr)
+            );
+            logger.info("Ran: {} {}", Arrays.toString(command), result);
+            return result;
+
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
             throw new RuntimeException(e);
         }
     }
 
+    private String readFileIfExists(Path path) throws IOException {
+        if (Files.exists(path)) {
+            long size = Files.size(path);
+            if (size > 100 * 1024) {
+                return "<<Too large to read: " + size  + " bytes>>";
+            }
+            LinkedList<String> result = new LinkedList<>();
+            AtomicBoolean linesDiscarded = new AtomicBoolean(false);
+            try(Stream<String> lines = Files.lines(path, StandardCharsets.UTF_8)) {
+                lines.forEach(line -> {
+                    result.add(line);
+                    if (result.size() >= TAIL_WHEN_TOO_MUCH_OUTPUT) {
```

Comment:
Why are we omitting lines at all? Once the VM is destroyed we can't get them out. If we want to debate about the merits of what is output fine, but I think it should be a separate PR from this already large PR to get the OS tests running in gcp?

Reply:
I'm going to remove omitting the lines, but will keep the size check.
One of the windows installer tests is generating a 16GB stderr.
I don't know what's in it because I didn't download it and nothing on windows could open it...